fog/view/connection/src/fog_view_router_client.rs

```diff
@@ -1,69 +1,236 @@
     ///
     /// Arguments:
     /// * uri: The Uri to connect to
+    /// * verifier: The attestation verifier
     /// * env: A grpc environment (thread pool) to use for this connection
     /// * logger: For logging
-    pub fn new(uri: FogViewRouterUri, env: Arc<Environment>, logger: Logger) -> Self {
+    pub fn new(
+        uri: FogViewRouterUri,
+        verifier: Verifier,
+        env: Arc<Environment>,
+        logger: Logger,
+    ) -> Self {
         let logger = logger.new(o!("mc.fog.view.router.uri" => uri.to_string()));
 
         let ch = ChannelBuilder::default_channel_builder(env).connect_to_uri(&uri, &logger);
-
         let fog_view_router_client = FogViewRouterApiClient::new(ch);
+        let (request_sender, response_receiver) = fog_view_router_client
+            .request()
+            .expect("Could not retrieve grpc sender and receiver.");
 
         Self {
-            fog_view_router_client,
             logger,
+            attest_cipher: None,
+            _fog_view_router_client: fog_view_router_client,
+            request_sender,
+            response_receiver,
+            uri,
+            verifier,
         }
     }
 
-    /// Makes streaming requests to the fog view router service.
-    pub async fn request(&self) -> Result<(), grpcio::Error> {
-        let (mut sink, mut receiver) = self.fog_view_router_client.request()?;
-        let attested_message = Message::new();
+    fn is_attested(&self) -> bool {
+        self.attest_cipher.is_some()
+    }
+
+    async fn attest(&mut self) -> Result<VerificationReport, Error> {
+        // If we have an existing attestation, nuke it.
+        self.deattest();
+
+        let mut csprng = McRng::default();
+
+        let initiator = Start::new(
+            self.uri
+                .responder_id()
+                .map_err(|_| Error::Attestation)?
+                .to_string(),
+        );
+
+        let init_input = ClientInitiate::<X25519, Aes256Gcm, Sha512>::default();
+        let (initiator, auth_request_output) = initiator
+            .try_next(&mut csprng, init_input)
+            .map_err(|_| Error::Attestation)?;
+
+        let attested_message: AuthMessage = auth_request_output.into();
         let mut request = FogViewRouterRequest::new();
-        request.set_query(attested_message);
-        let send = async move {
-            for i in 0..5 {
-                log::info!(self.logger, "Sending message {}", i);
-                sink.send((request.clone(), grpcio::WriteFlags::default()))
-                    .await?;
-            }
-            sink.close().await?;
-            Ok(()) as Result<(), grpcio::Error>
+        request.set_auth(attested_message);
+        self.request_sender
+            .send((request.clone(), grpcio::WriteFlags::default()))
+            .await?;
+
+        let mut response = self
+            .response_receiver
+            .try_next()
+            .await?
+            .expect("Auth response was not received.");
```

Comment:
This probably shouldn't be an `expect()`---like, it's basically "if the specific fog URL I'm trying to talk to doesn't get back to me, crash the app"

Reply:
See comment below. This client is really just for tests so this isn't a true network call. That said, I'll remove it.